### DISCUSSION TITLE
Curated Extensibility Principles

### DISCUSSION THREAD

**Discussion by alexflorisca** [woocommerce/woocommerce#41304] [General]
These are some of the thoughts that capture the extensibility guidelines for WooCommerce. They should be followed by all code contributors, and serve as a basis for decisions around new extensibility points. For now, it's a working document and I would love some input if and how these apply across the whole of WooCommerce.

## Favor explicit interfaces for integrations over action and filter hooks

Wherever possible, instead of using do_action or apply_filters (or the equivalent @wordpress/hooks interfaces in the client) – we introduce explicit interfaces for adding extensibility. Some examples are the [payment method integration interfaces](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/checkout-payment-methods/payment-method-integration.md?) which include client side registration functions[registerBlockComponent](<https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/blocks-registry/block-components/register-block-component.js#L69>).

This allows for:
- Typing and clear expectations around arguments and configuration.
- Hiding implementation details from consumers (so more flexibility in changing implementation as long as the public interface contract doesn’t change).
- Clearer expectations used in covering the functionality in automated tests.
- More trivial deprecation of interfaces in the future if the extensibility needs change

## Where action and filter hooks are needed, keep the surface area of impact as small as possible

Of course, being that we are building in a WordPress context, it’s near impossible to avoid the usage of PHP action and filter hooks as that is a primary mechanism used server-side for extensibility. It is battle tested and proven.

As much as possible:
- Keep the list of exposed arguments as small as possible. We won’t add any additional arguments because we think plugins might need them in the future.
- For filters, we are careful about passing in additional arguments that could be mutated by plugins unexpectedly and cause unexpected side-effects.

## Prefer names for APIs/Interfaces that are not layout specific.

For example, including names like Sidebar or Footer or Header in extensibility APIs couples the interface to layout. Layout is something that is volatile especially if the interface/API is in some UI that changes between desktop, mobile, or apps.

## Encapsulate individual components and protect against catastrophic errors

One developer mistake shouldn’t punish another developer. Merchants deserve to know which plugin isn’t working, and the shopper/merchant experience (especially during the checkout flow) shouldn’t be crippled because a single plugin didn’t work well.
Plugins should be encapsulated individually, meaning one plugin failing in a filter shouldn’t break other plugins integrating into that filter or other extensibility interface.. 

There are numerous techniques that are implemented here:
 - Ensure filtered values are of the correct type and shape before using (and fallback to defaults).
 - Validate plugin configuration arguments on interfaces (type checking etc).
 - Surfacing errors in the appropriate contexts (eg. consider surfacing errors more visibly to administrators and silently logging for visitors).
 - Implement error boundaries where possible (eg. React error boundaries or try & catch).

## Plugins cannot change the data structure
It’s important that core experience is preserved. Having the shape and predictability of data fixed, avoids breaking the core experience, and avoids breaking third party code that depends on that data

This is another important principle that relates to stability. Plugins shouldn’t edit the core experience (whenever possible), they enhance and add to it, by only adding things, not removing them. For example, through slot fills or registering custom blocks, other plugins can add UI elements to the Cart or Checkout block, but they cannot remove any.

This means plugins:
- Should not edit the Store API response
- Should not have access to hooks that edit context

## Only expose what’s necessary
Anything exposed for public use creates a contract establishing a burden that needs maintenance. It immediately sets constraints on how that contract could be changed in the future, should conditions change for feature/interface requirements.

Dedicated public APIs are easier to maintain and reason about because future engineers can quickly know that those are public and not internal.

When exposing any data to other plugins, be very intentional about what is exposed. It might be helpful to expose a lot of data to help with an integration, for example. This increases the public contract and the chance of backward compatibility breaking. Instead, things that already have a public contract should take priority, like the return value of Store API or dedicated public API proxies that consume internal data and reshape it for external use.

**Comment by jonathan-dejong** [comment]
It's nice to see this becoming an official set of guidelines! 

I've got a couple of comments 😃 

> For filters, we are careful about passing in additional arguments that could be mutated by plugins unexpectedly and cause unexpected side-effects.

Am I misunderstanding something here since the additional arguments passed to `apply_filters()` and in extension to the callback of a `add_filter` is not mutable to begin with. Unless you were to pass them as references which I only know a handful of examples of in WP core (for example the `pre_get_posts` action hook). 

Only the first argument is the one affected by `apply_filters()` so why not just ensure that you pass as much useful additional information as possible with the filters? 
Note: I assume we're talking about PHP hooks here since that's set as the premise just above:
> Of course, being that we are building in a WordPress context, it’s near impossible to avoid the usage of PHP action and filter hooks as that is a primary mechanism used server-side for extensibility. It is battle tested and proven.


My second feedback is in the same vein regarding:
> Keep the list of exposed arguments as small as possible. We won’t add any additional arguments because we think plugins might need them in the future.

Again I question the stance of "as small as possible" vs an approach like "as useful as possible". 
I also don't see why it'd be so horrible to add additional arguments to existing hooks down the line. Since there is no breaking behaviour in doing so with either `apply_filters()` or `do_action()`. 
If I as a "consumer" of the hooks don't add a specific number of arguments in my callback, it assumes 1. If I add a specific number it only expects that as a minimum. 

For example if I do `add_filter( 'filter_name', 'filter_callback', 10, 2 );` and you add a third argument, this won't break my code. I'll just not be using your additional argument. 
A classic example of this is the `save_post` hook. https://developer.wordpress.org/reference/hooks/save_post/
This passes three arguments; `$post_id, $post, $updated` but do a search online and you'll find tons and tons of example code where only the first argument is used in the callback. Few uses `$post` and barely anyone seems to use `$updated` in my experience. 

All of this ties in with a long standing grievance I've had with the WC core codebase for many years now where I've been wanting to use a particular filter or action but then the arguments provided through the hook is not enough to make it useful to me. 
Then upon examining the code I can see that the information is there, it's just never passed to the hook. 
_I took a few minutes to try to find a good example of this but I've been working with a non-WC project now for too long so I couldn't remember one at the time of writing this._ 

-------

Lastly regarding this:
> This is another important principle that relates to stability. Plugins shouldn’t edit the core experience (whenever possible), they enhance and add to it, by only adding things, not removing them. For example, through slot fills or registering custom blocks, other plugins can add UI elements to the Cart or Checkout block, but they cannot remove any.

I understand the reasoning behind this, but it is a rather big crutch on customising the experience of a WC shop. 
As an example, a lot of our clients nowadays are requesting a distraction free checkout experience for their customers. This often means that we need to modify and remove information that WC adds by default to the checkout that just isn't applicable to the clients shop. 
So will this be a problem if this is the stance WC takes with its code? Or will you then adopt a much higher standard of editability where most things can be toggled on/off through block settings (for example)? 

A typical such request is for example to remove information from the "order review table" in favour of showing the cart above the checkout. 

------
I look forward to your response. As previously stated I think having this type of guideline is great! 👍

**Reply by senadir** [reply]
> I understand the reasoning behind this, but it is a rather big crutch on customising the experience of a WC shop.
As an example, a lot of our clients nowadays are requesting a distraction free checkout experience for their customers. This often means that we need to modify and remove information that WC adds by default to the checkout that just isn't applicable to the clients shop.
So will this be a problem if this is the stance WC takes with its code? Or will you then adopt a much higher standard of editability where most things can be toggled on/off through block settings (for example)?

I don't think this rule is something we're going for enforce on everything. We're going to differentiate between things that are expected to be deleted and things fundamental to the flow.

To give you an example. You can customize the Checkout template however you want to get a distraction free experience, but for example, we will not allow deleting the country field or the email field, because they're a fundamental part of the flow and many plugins and core flows depend on them existing.

We also won't allow deleting the shipping address for a cart that requires shipping and/or has a selected shipping rate.

So our approach isn't 100% extensibility, which might be too powerful for its own good, and is something that puts developers at odds with each other.

When we ensure you can edit what's expected to be edited, and add upon, all developers can expect a consistent experience. You should not for example, always assume a phone number is going to be present unless you somehow force it (which is doable). The same way that you can always expect there to be a payment method for an order with sum above 0.

**Reply by senadir** [reply]
When it comes to filters, another concern that we take early on is that the more arguments we pass, the larger the surface area we're exposing ourself to. Having more parameters means we're unlikely to refactor the code if that extra parameter is no longer part of the code, as we assume the majority of stuff in the flow are implementation details.

You shared an example of using just the first argument, but what if you're using 4 and we reduce them to 3 because the 4th one no longer exists in that flow.

When we pass an object as a the primary arg for a filter or an object, we actually expect you to edit that than to fetch one from DB and edit it and save it.

However, if you see a filter in Checkout Block that you think is missing arguments, do let us know, when it make sense, and if we think that argument is not part of the internal implementation detail, we will expose it.

**Comment by rrennick** [comment]
> Unless you were to pass them as references which I only know a handful of examples of in WP core (for example the pre_get_posts action hook).

@jonathan-dejong Since PHP 5.0 objects are [passed by reference by default](https://www.php.net/manual/en/functions.arguments.php#119724).

**Reply by jonathan-dejong** [reply]
Right, so that's one data type that the concern legitimately applies to.
I do not know of any other, do you? 

So the original statement would still be incorrect for any passed data type apart from objects, which can in that case be it's own guideline like "Minimize use of additional objects in hooks" or something like that :)

**Reply by nerrad** [reply]
@jonathan-dejong I don't think the existing statement is contradictory here. It's also not dogmatic but simply highlights that care should be given to what is exposed as an argument on the action or filter. 

@alexflorisca, @senadir I wonder if we should remove the PHP specificity in this particular principle being discussed since WordPress also has [JS based action and filter hooks](https://github.com/WordPress/gutenberg/tree/trunk/packages/hooks) that the principle applies to.

**Comment by NTShop** [comment]
Your link for "payment method integration interfaces " is broken. Maybe this is the link: 

https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/checkout-payment-methods/payment-method-integration.md

**Reply by alexflorisca** [reply]
I've updated both links in the original post, thanks for posting this

**Comment by nerrad** [comment]
One thing that the current proposed principles don't call out explicitly, but probably should, is something along the lines of favoring Merchant/Shopper experience over Dev Ex. In otherwords, any exposed extensibility surfaces should enable/support great UX for merchants and shoppers. 

I realize there's a tension there because truly flexible interfaces will always carry with them a degree of risk of abuse that break UX. 

I do think this is something we should wrestle with though.

**Reply by opr** [reply]
Could you clarify your thoughts a little here @nerrad, I'm not sure _exactly_ what your suggestion is trying to solve.

Are you suggesting that it's OK for extensibility interfaces to be more difficult for developers to use or require extra steps to achieve something if making them easier/more flexible could allow extensions to degrade the shopper UX? If so, I agree, if not please can you help me understand better?

> In otherwords, any exposed extensibility surfaces should enable/support great UX for merchants and shoppers.

Can you think of any instances where they _don't_ do this?

**Reply by nerrad** [reply]
The discussion opens with this statement:

> They should be followed by all code contributors, and serve as a basis for decisions around new extensibility points.

The first question that comes to mind is "why?". Why is it important that they should be followed by all code contributors and serve as a basis for decisions? Each principle has some rationale for the principle, but what is the overarching reason for having principles to begin with?

I'm suggesting that these principles should be anchored in enabling the creation and maintenance of a great merchant and shopper experience. Developer experience should always serve that primary goal.

**Reply by nerrad** [reply]
I would also say that I do believe these principles are developed with that mindset, but we haven't explicitly called that out in the proposal.

**Reply by opr** [reply]
OK thanks Darren, I've been thinking about how to reword this since last week, it's pretty tough.

Originally we had:

> These are some of the thoughts that capture the extensibility guidelines for WooCommerce. They should be followed by all code contributors, and serve as a basis for decisions around new extensibility points. For now, it's a working document and I would love some input if and how these apply across the whole of WooCommerce.

I propose we could insert an introduction to the guidelines like this.

> This is a set of principles to follow when contributing code to WooCommerce. They are written to guide contributors and have the ultimate goal of ensuring anything we add to the codebase will improve merchant and shopper experiences.
>
> Extensions should integrate smoothly with WooCommerce to enhance the user experience, and the extensibility points they use should enable them to do this.
>
> Code contributors should follow these principles so that all contributions are part of a collective effort to improve the overall user experience. These guidelines serve as a basis for decisions on new extensibility points. For now, it's a working document and I would love some input if and how these apply across the whole of WooCommerce.

Disclaimer: I originally wrote a paragraph of text and asked GPT to improve what I'd done. I then did some heavy editing on the output. In spite of this, I am still finding it tough to get something that really encapsulates what the goal here is, but feel this is closer.

**Comment by NTShop** [comment]
I'm not clear on something, who decides what will "degrade the shopper UX" and what are the standards that are used to determine what the shopper UX should be like? Are a handful of WC core developers the people to decide, or is it the many almost innumerable merchants who prefer certain ways of doing things based on their own industry experience and their own customer feedback? And if it's the latter shouldn't it be as easy as possible for the UX to be modify to meet their preferences?

**Reply by senadir** [reply]
>  Are a handful of WC core developers the people to decide, or is it the many almost innumerable merchants who prefer certain ways of doing things based on their own industry experience and their own customer feedback?

I think that's a false dichotomy to base the argument on. Think of WooCommerce how you would think of Shopify, Wix, and any other SaaS. Do you look at those solutions and proclaim that a handful of developers are calling the shots? WooCommerce isn't run by a handful of developers, it's very much a company, who happens to base to have open source as its DNA, and each company employees a variety of people.

Design decisions are done by experienced industry designers and product managers, who would take several variables, including existing research, competitive overview, data, merchant interviews, and developers feedback, very much like yours. Which would be viewed and referenced.

Once such decisions are made, we, developers, are tasked with taking such design to be materialized into the core product.

 As when it comes to giving merchants all the power, that has proven to give bad results constantly for the majority of merchants. We strive to always listen to the problems you're facing, the customizations you're trying to do, and the outcome you're looking for, and how we can either build it into WooCommerce, or provide the tools for you to build.

**Reply by jonathan-dejong** [reply]
It's easy to forget that WordPress and WooCommerce is run as "benevolent dictatorships" 😅 
(https://en.wikipedia.org/wiki/Benevolent_dictator_for_life)

This comment from @NTShop is in a similar vein to a comment I made here:
https://github.com/woocommerce/woocommerce/issues/42057#issuecomment-1850734744
> Either way, this means that WooCommerce thinks it should dictate what a good checkout UX is for all its user base?

and while I don't think either you @senadir nor @pmcpinto have given unsatisfactory responses it may be worth considering why this keeps coming up in your discussions with the community? 

Maybe WooCommerce needs to be more transparent about its decision making process? When announcing a roadmap, new feature, a decision, also back that up with the _"why"_. 

Why did you choose to ship the checkout and cart block with the current features? 
Why did you decide to not make it possible to do x/y/z? 
Why did you start with HPOS and not HPPS (high performant product storage)? 
Why did you decide not to take the opportunity to finally add a dedicated order_number value with HPOS (yes this one still bugs me)?

A lot of the time it's not that I disagree with what you're doing, or that I think it's bad, wrong, {insert negative adjective}. A lot of the time it's mostly frustration over having my future tooling as a developer using WooCommerce dictated for me without understanding the _why_.  

I realise I maybe made this a smidgen OT, apologies :D I just read through all the comments on these recent discussions and this is where I found my train of thought to end.

**Reply by pmcpinto** [reply]
Just to complement what was mentioned by Nadir:

- Initially, the Cart and Checkout blocks were designed based on industry best practices, considering qualitative and quantitative data.
- We plan to keep improving the experience, as having an optimized cart and checkout is critical for any business operating in a highly competitive environment. We’re also exploring ways to showcase how specific changes can enhance the experience.
- We recognize that there’s no one-size-fits-all solution. This is why we plan to invest on [custom Checkout fields](https://github.com/woocommerce/woocommerce-blocks/discussions/11173#discussioncomment-7403117), [custom content on the Cart/Checkout summary](https://developer.woo.com/roadmap/custom-content-on-the-cart-and-checkout-summaries/), and more initiatives we have in the pipeline.
- Like any organization, our company is a complex adaptive system with limited resources. Prioritization and trade-offs are inevitable due to the dynamic nature of our environment and depend on the resources available to work on the Checkout surface areas.

> is it the many almost innumerable merchants who prefer certain ways of doing things based on their own industry experience and their own customer feedback?

We’re always eager to learn more about how merchants make decisions. Would you mind sharing more about this topic:
What are the primary use cases that your clients aim to address?
How do your clients collect feedback and validate that the implemented changes positively impact the overall user experience? 

Thanks in advance!

**Comment by NTShop** [comment]
> As when it comes to giving merchants all the power, that has proven to give bad results constantly for the majority of merchants.

Are you saying that in reference to the majority of untold millions of WC sites? Wow. Is that data and data gathering methodology available for review somewhere? It'd be very interesting so that see that info. 

My concern is whether or not we 3rd party developers for the plethora of WC sites need to adjust our info so as to be able to tell customers what is and isn't possible when they ask about this or that type of functionality.

**Comment by senadir** [comment]
I wrote an initial take at this principle

### Favoring PHP extensibility over JS 

Building extensibility for WooCommerce will always require PHP, but not JS. Doing a single task in JS that would require a build process, enqueue scripts and managing dependencies is tricky, so when possible, start with a PHP extensibility API. If the API is something you expect to change often in a Checkout life cycle, then it should probably be JS, for most other things like options, settings, things that affect the initial state of Checkout, and fully controlled APIs, it's preferable to have a single PHP extensibility API that manages that.

**Reply by jonathan-dejong** [reply]
I like this! Makes perfect sense to me and it's one of the things I find confusing with the gutenberg project right now.. there's a PHP and JS approach to a lot of the same things and it's tricky knowing the pros/cons between each approach or which is going to be favoured going forward in WP core. 

If WC can keep to one approach for each "thing" that makes our lives easier :)

**Reply by nerrad** [reply]
I like the direction of this. One suggestion I would make is to remove the explicit reference to Checkout as I think this principle should apply to extensibility in general within Woo. You could use Checkout as an _example_ of applying the principle.

I think another nuance we'll see with PHP vs JS extensibility is that in some cases, extensibility interfaces will be available in both contexts - but the JS context allows for more bespoke and/or complex _interactive_ customization.
